src/coord/src/coord.rs

```diff
@@ -2207,6 +2289,44 @@ impl Coordinator {
         mut session: Session,
         action: EndTransactionAction,
     ) {
+        // If we are trying to commit a write transaction but another connection has a
+        // write in progress, queue ourselves.
+        if EndTransactionAction::Commit == action
+            && self.write_inprogress != None
+            && self.write_inprogress != Some(session.conn_id())
+        {
+            let txn = session
+                .transaction()
+                .inner()
+                .expect("must be in a transaction");
+            if let Transaction {
+                ops: TransactionOps::Writes(_),
+                ..
+            } = txn
+            {
+                self.serialized_writes.push_back(PlanReady {
+                    tx,
+                    session,
+                    plan: Plan::CommitTransaction,
+                });
+                return;
+            }
+        }
+
+        // If this transaction blocked other writes, allow them to proceed.
```

Comment:
I don't think this matters in practice due to the coord's single-threaded nature, but "ethically" I was surprised to see this happen before `sequence_end_transaction_inner`. Like this is saying we can start other writes before executing the effects of this transaction from my read. Maybe this is just a useless nit, so feel free to disregard! If my read of this is entirely wrong, though, would love to have my thinking straightened out.

Reply:
I completely agree, it could/should be moved down.